homeassistant/components/iotawatt/sensor.py

```diff
@@ -169,6 +169,10 @@ def __init__(
             self._attr_unique_id = (
                 f"{data.hub_mac_address}-input-{data.getChannel()}-{data.getUnit()}"
             )
+        elif data.getType() == "Output":
+            self._attr_unique_id = (
+                f"{data.hub_mac_address}-output-{data.getName()}-{data.getUnit()}"
```

Comment:
As per the linked documentation above by @agners above, the markers used in this unique ID don't meet the requirements the Home Assistant project has set.

Going to close this PR for that reason. Nevertheless, thanks for being willing to contribute 👍 

../Frenck

Reply:
Can you please clarify exactly what the concern is here?  The docs state:
> It is important that it is not possible for the user to change the unique ID, because the system would lose all its settings related to the unique ID.

Changing the name of an Iotawatt output would be very rare, only needed when changing the house electrical system.  In this case, losing all its settings and treating it as a new entity is precisely what I would expect to happen.
On the other hand, the benefit of this PR would be seen by every user because the entities will be grouped under the device.

Reply:
+1

Simply closing this PR seems rather heavy handed. Every user of the iota watt want this, why deny them of this feature?

Reply:
Our docs and policies are unique IDs are very clear about the requirements. There is no discussion possible on if this may be exempted or possible or not.

Sorry.

../Frenck

Reply:
Open source software thrives on free exchange of ideas, discussion, and debate. Your statement of “there is no discussion possible” is highly disappointing.

Reply:
This debate has been done a couple of times before, the rules exists because of past experiences and debates. Which is why it is pretty explicit and extensive documented.

Your above comment is therefore a bit odd and out of place.

It has bitten us before. As per above, we are happy to accept a PR that meets the requirements this project has set.

Thanks 👍 

../Frenck

Reply:
> This debate has been done a couple of times before, the rules exists because of past experiences and debates. Which is why it is pretty explicit and extensive documented.
> 
> Your above comment is therefore a bit odd and out of place.
> 
> It has bitten us before. As per above, we are happy to accept a PR that meets the requirements this project has set.

For all usage and purpose, data.getName() is a unique identification as far as the iotawatt works.
It will last for the lifetime of the iotawatt. If it changes, it's an extremely high possibilities that it's actually measuring something new which is exactly what a unique name is for: it should and will be a new entity.

Stating it is not and dismissing the PR on the basis of "requirements" is what I think is odd.

You have provided zero explanation on why it would be a bad idea for the iotawatt integration, for the both times you rejected it.
Every single time the answer has been the same "don't meet the requirements".

Reply:
> If it changes, it's an extremely high possibilities that it's actually measuring something new which is exactly what a unique name is for: it should and will be a new entity.

It is changeable and this does not provide a unique ID that meet our requirements.

Home Assistant needs a constant and unique reference to keep the registries, automations, history and many more in sync. It is not just entities.

If an integration cannot provide an unique ID, it is totally fine to omit it.

../Frenck

Reply:
If it changes, it is no longer the same output. On that premise, the name configured in the iotawatt is a unique ID.
Again, for all purposes and usage: the name set in the iotawatt is unique.

If an output was set to be "Solar Output, and the user now wants it to be "dishwasher" why would you ever want to consider those one and the same and preserve history from one to the next. It's nonsensical.

Ultimately, the issue isn't that the iotawatt's output sensor doesn't have a unique ID per say. Is that in the UI, they can't be easily given a friend name, can't be modified or interacted with the way an input sensor is.

Reply:
FWIW, I created a feature request in the IoTAWatt firmware: https://community.iotawatt.com/t/feature-request-unique-identification-of-outputs/5444

Reply:
> If it changes, it is no longer the same output. On that premise, the name configured in the iotawatt is a unique ID.

I get that response, however, the same could be said for changing the name of a Philips light bulb when moving it to a different room (or any other name of any other device). Yet, it is user changeable, eg not because of a change in use; but naming changes changes in strategies (eg prefixing names, correcting typos, adjusting meta data in a name), are not by definition functional change to an datapoint values.

Reply:
> This is a great example. So let say I do have "Solar Output" and now maybe install a second solar: Solar Output West. It I now change the name of the first for clarity to "Solar Output East", things go south.

In the end, for the user this is a *very* *very* small caveat compared to the downsides of not having all the benefits of what an entity gets with unique ID these days (e.g. display precision :heart_eyes: ). 

In IoTaWatt, the output name *is* actually the ID: This can be seen if you press "add", type "Solar", enter a formula and save, then press "add" again, type "Solar" again, add a different formula, the first entity gets overwritten :sweat_smile: . This is not ideal user experience, and if it would have a proper generated unique id, this would not be the case. But in the end, that is how it works currently :man_shrugging:  Luckily, IoTAWatt is open source, so maybe we can get a proper unique id for outputs and have the best user experience. Let's see.

However, IMHO, we should allow user configured "ids" for entities in such corner cases. Formally, I'd say these conditions would need to be met:

1. the higher level device has a proper unique id
2. the entity meaning itself is user configurable on the external system (e.g. a group or a calculated value like the issue at hand)
3. the id is unique (as in, there is no way that the same "configured id" will appear twice)
4. there is no alternative